src/fairness/print_hierarchy/flux_shares.cpp

```diff
@@ -0,0 +1,156 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+#include <iostream>
+#include <iomanip>
+
+extern "C" {
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+}
+
+#include "src/fairness/reader/data_reader_db.hpp"
+
+using namespace Flux::accounting;
+using namespace Flux::reader;
+
+const std::string DBPATH = std::string (X_LOCALSTATEDIR) + "/FluxAccounting.db";
+
+static void show_usage ()
+{
+    std::cout << "usage: flux shares [-P] [DB_PATH]\n"
+              << "optional arguments:\n"
+              << "\t-h,--help\t\tShow this help message\n"
+              << "\t-P,\t\t\tPrint the database hierarchy in a parsable format"
+              << std::endl;
+}
+
+std::stringstream read_from_db (const std::string &filename)
+{
+    data_reader_db_t data_reader;
+    std::shared_ptr<weighted_tree_node_t> root;
+    std::ostringstream out;
+    std::stringstream buffer;
+
+    root = data_reader.load_accounting_db (filename);
+    if (root == nullptr)
+        return buffer;
+
+    weighted_walk_t walker (root);
+
+    walker.dprint_csv (out);
+    buffer << out.str();
+
+    return buffer;
+}
+
+
+/*
+print_hierarchy () takes two arguments: a filepath to a flux-accounting DB, and
+a boolean value 'parsable'. It will call load_accounting_db () to get a
+hierarchy of users. It will then parse each line, fetching tokens delimited by
+','. The formatting for lines will be different if a line contains a bank or if
+it contains an association.
+
+If parsable is false, the output of the database hierarchy will be
+formatted into columns.
+*/
+int print_hierarchy (const std::string &filename, bool parsable)
```

Comment:
Rather than parsing the output of the `dprint_csv` walker method, can you implement your own walker method?  Something like `print_csv` to print the parsable format (with an optional argument for the delimiter) and `print_pretty` for the human readable version?

Reply:
I think I've got a solution. After some fiddling around, I ended up with two different functions: `print_csv ()` and `pretty_print ()`. Both of these start at the root node of a weighted tree and traverses it, printing all children. It should traverse the tree recursively until it reaches all leaf nodes. 

`print_csv ()` can take a custom delimiter if one is passed in via the `-P` option. I added a test to demonstrate this with `,`:

```
Account,Username,RawShares,RawUsage
root,,1000,133
 account1,,1000,120
  account1,leaf.1.1,10000,100
  account1,leaf.1.2,1000,10
  account1,leaf.1.3,100000,10
 account2,,100,12
  account2,leaf.2.1,10000,10
  account2,leaf.2.2,1000,1
  account2,leaf.2.3,100000,1
 account3,,10,1
  account3,leaf.3.1,100,0
  account3,leaf.3.2,10,1
```